src/thunderbots/software/util/gradient_descent.h

```diff
@@ -0,0 +1,180 @@
+         * @param past_gradient_decay_rate Past gradient knowledge decay rate, see
+         *                                 corresponding class member variable for details
+         * @param past_squared_gradient_decay_rate Past squared gradient knowledge decay
+         *                                         rate, see corresponding class member
+         *                                         variable for details
+         */
+        GradientDescentOptimizer(ParamArray param_weights,
+                                 double gradient_approx_step_size,
+                                 double past_gradient_decay_rate,
+                                 double past_squared_gradient_decay_rate);
+
+        /**
+         * Attempts to maximize the given objective function
+         *
+         * Runs gradient descent, starting from the given initial_value and running for
+         * num_iters
+         *
+         * @param objective_function The function to maximize
+         * @param initial_value The value to start from
+         * @param num_iters The number of iterations to run for
+         *
+         * @return The parameters corresponding to the maximum value of the objective
+         *         found
+         */
+        ParamArray maximize(std::function<double(ParamArray)> objective_function,
+                            ParamArray initial_value, unsigned int num_iters);
+
+        /**
+         * Attempts to minimize the given objective function
+         *
+         * Runs gradient descent, starting from the given initial_value and running for
+         * num_iters
+         *
+         * @param objective_function The function to minimize
+         * @param initial_value The value to start from
+         * @param num_iters The number of iterations to run for
+         *
+         * @return The parameters corresponding to the minimum value of the objective
+         *         found
+         */
+        ParamArray minimize(std::function<double(ParamArray)> objective_function,
+                            ParamArray initial_value, unsigned int num_iters);
+
+        /**
+         * Attempts to minimize or maximize the given objective function
+         *
+         * Runs gradient descent, starting from the given initial_value and running for
+         * num_iters
+         *
+         * @param objective_function The function to minimize
+         * @param initial_value The value to start from
+         * @param num_iters The number of iterations to run for
+         * @param gradient_movement_func The function to use on each step along the
+         *                               gradient, either "-" to minimize the given
+         *                               function, or "+" to maximize it
+         *
+         * @return The parameters corresponding to the minimum or maximum value of the
+         *         objective found, depending on what gradient_movement_func was given
+         */
+        ParamArray followGradient(
```

Comment:
nit: Seems like this function is only for internal use and shouldn't be exposed publicly. Consider moving it to `private`?

Reply:
Oh absolutely, I thought I had done that. For the record, I would consider this to be much more then a nit, public class interfaces are super important